### ISSUE TITLE
e10cfdd8953 regresses hf_Longformer-cuda-eager on pytorch 2024-01-16

### ISSUE THREAD

**Issue by xuzhao9** [pytorch/pytorch#118548]
### 🐛 Describe the bug

[e10cfdd8953](https://github.com/pytorch/pytorch/commit/e10cfdd89531366c5dd0b8dfa3313b2d4be06457) regresses  hf_Longformer-cuda-eager on pytorch 2024-01-16:

```
{
  "target_repo": "pytorch",
  "start": "8bcdde5058658cc193c94a7f1eb16660553dc35a",
  "start_version": "2.3.0.dev20240110+cu118",
  "end": "28bb31e4a588f3fa198e927439fe09c871201aab",
  "end_version": "2.3.0.dev20240116+cu118",
  "result": [
    {
      "commit1": "7e6a04e542f",
      "commit1_time": "2024-01-10 22:25:40 +0000",
      "commit1_digest": {
        "name": "torch-nightly",
        "environ": {
          "pytorch_git_version": "7e6a04e542ff4b1cd87dc2dba9095274604b26af",
          "pytorch_version": "2.3.0a0+git7e6a04e",
          "device": "NVIDIA A100-SXM4-40GB",
          "git_commit_hash": "7e6a04e542ff4b1cd87dc2dba9095274604b26af"
        },
        "metrics": {
          "test_eval[hf_Longformer-cuda-eager]_latency": 142.44164,
          "test_eval[hf_Longformer-cuda-eager]_cmem": 0.7841796875,
          "test_eval[hf_Longformer-cuda-eager]_gmem": 5.22845458984375
        }
      },
      "commit2": "e10cfdd8953",
      "commit2_time": "2024-01-10 23:16:42 +0000",
      "commit2_digest": {
        "name": "torch-nightly",
        "environ": {
          "pytorch_git_version": "e10cfdd89531366c5dd0b8dfa3313b2d4be06457",
          "pytorch_version": "2.3.0a0+gite10cfdd",
          "device": "NVIDIA A100-SXM4-40GB",
          "git_commit_hash": "e10cfdd89531366c5dd0b8dfa3313b2d4be06457"
        },
        "metrics": {
          "test_eval[hf_Longformer-cuda-eager]_latency": 183.184652,
          "test_eval[hf_Longformer-cuda-eager]_cmem": 0.7841796875,
          "test_eval[hf_Longformer-cuda-eager]_gmem": 5.23040771484375
        }
      }
    }
  ]
}
```

Before latency: 142.44 ms
After latency: 183.18 ms

### Versions

https://github.com/pytorch/pytorch/pull/117067

cc @ezyang @gchanan @zou3519 @kadeng @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by ezyang** [comment]
Cc @soulitzer

**Comment by soulitzer** [comment]
For the 2D x ND / ND x 2D case, `_matmul_impl` which is composite can dispatch to `mm` or `bmm` and that determination is via the following logic:
  - If ND can be flattened into 2D without copy:
    - mm (case 1)
  - else:
    - if we need to compute grad wrt 2D tensor
      - mm (case 2)
    - else
      - bmm (case 3)

The reason we want to do mm even if flattening in the 2D tensors incurs copies in case 2 is that doing this helps us avoid materializing large intermediates:

If we did bmm, the forward looks like:
`[b, n, k] @ [k, m] = [b, n, m]`
And if  [k, m] requires grad, to compute its gradient, we need to materialize a intermediate which subsequently reduce away during backward
`sum([b, n, k].T @ [b, n, m]) => sum([b, k, m]) => [k, m]`

In the original issue that the tagged PR is fixing, someone is doing `out=` with tensors that require grad but inside no-grad region. 

We expect this to fall into case 3 because (1) 2D x ND (as opposed to ND x 2D) can NEVER be done without copy and (2) out= does not work with autograd. However, because the user actually passed in tensors that required grad, even though we are in no-grad, this tricked the logic into believing that we fall into case 2. This led to an internal assert.

The fix for the original PR is to improve the logic so that the user's case correctly falls into case 4. The was the intention of the original logic as well, so this is a purely a bug fix.

However, the side effect of that PR is that more cases in general move from case 3 to 2. It's possible that in this specific case, the mm kernel so much faster than the bmm kernel such that it is still desirable to call into mm even if it requires additional copying.

More investigation is necessary to how to make the mm + copies vs bmm trade off. Should matmul try to be smart and make a decision based on information about the backend, or perhaps we can exposing a knob like `force_fold_and_use_mm` that users can tune themselves? I don't have an opinion on this right now (maybe others do?).

An easy way out here that does not having develop any deeper understand here is to revert this PR and try to fix the originally reported issue in a more narrow way i.e., add the condition "don't go into mm path even if 2D requires grad and using out=". This avoids the side effect while not regressing the fix for the issue either.

**Comment by lezcano** [comment]
Revert sgtm. As to how to approach this issue, on eager there is so much we can do, but this is something we can certainly tune in `torch.compile` with max-autotune.  cc @eellison @shunting314

**Comment by xuzhao9** [comment]
@soulitzer Can we have the concrete input shapes of bmm and mm kernels specifically in this model? We are also interested in looking into why bmm kernel is so much slower than torch.mm counterparts.

**Comment by xuzhao9** [comment]
I did some deeper analysis. Before this PR, nn.Linear uses `aten::mm`:
<img width="812" alt="image" src="https://github.com/pytorch/pytorch/assets/502017/cb1d3561-ac8c-485c-8fe4-662e5c5cea64">

One of the `nn.Linear` takes 0.108 ms walltime.

After this PR, nn.Linear uses `aten::bmm`:
<img width="812" alt="image" src="https://github.com/pytorch/pytorch/assets/502017/4795d853-eca3-4e8f-9f90-7be3d52400bf">

The same `nn.Linear` takes 0.142 ms walltime, this is 40% slowdown.

**Comment by soulitzer** [comment]
Thanks for looking into that. I was actually unable to repro using `pytest test_bench.py  -k "test_eval[hf_Longformer-cuda]"`.

I did observe a significant difference in runtime between the bmm and mm paths (on A100). 
Some more results on some more input sizes in the form of `[768, 768] @ [B, 768, M]` still with fp16 (runtime in ms):

```
B          M            mm + copy          bmm
---------------------------------------------------
2048       4            0.152385           1.432394
1024       8            0.147818           0.488963
512        64           0.692234           0.352137
128        128          0.390359           0.129712
64         512          0.773324           0.221456
8          1024         0.228340           0.089903
4          2048         0.229695           0.090430
```

<details>

<summary>
click for repro
</summary>

```python
import torch
import time

from torch.testing._internal.logging_tensor import LoggingTensor, LoggingTensorReentrant, LoggingTensorMode, \
    log_input, capture_logs, capture_logs_with_logging_tensor_mode

def print_table_header():
    print(f"{'b':<10} {'m':<12} {'mm + copy':<18} {'bmm':<18}")

def print_table_row(b, m, time_grad_true, time_grad_false):
    print(f"{b:<10} {m:<12} {time_grad_true:<18.6f} {time_grad_false:<18.6f}")

def measure(B, M):
    def run(x, y):
        out = x @ y
        torch.cuda.synchronize()
        return out

    # [n, k]
    a = torch.rand([768, 768], dtype=torch.half, device="cuda", requires_grad=True)
    a_detach = a.detach()

    # [b, m, n]
    b = torch.rand([B, M, 768], dtype=torch.half, device="cuda").transpose(1, 2).contiguous()

    def fn(t1, t2):
        return t1 @ t2

#    compiled_fn = torch.compile(fn, mode="max-autotune")
#    compiled_fn(a_detach, b)
#
#    def wrapped(t1, t2):
#       fn(t1, t2)
#       torch.cuda.synchronize()
#
#    # warm
#    for _ in range(1000):
#        wrapped(a_detach, b)
#
#    before = time.perf_counter()
#
#    # test
#    for _ in range(1000):
#        wrapped(a_detach, b)
#    after = time.perf_counter()
#    print(after - before)


    # print("require_grad=True")
    # warm
    for _ in range(1000):
        run(a, b)

    with capture_logs_with_logging_tensor_mode() as logs:
       run(a, b)
    # print("\n".join(logs))

    before = time.perf_counter()

    # test
    for _ in range(1000):
        run(a, b)
    after = time.perf_counter()
    t_req_grad_true = after - before


    # print("require_grad=False")
    # warm
    for _ in range(1000):
        run(a_detach, b)

    with capture_logs_with_logging_tensor_mode() as logs:
       run(a_detach, b)
    # print("\n".join(logs))
    before = time.perf_counter()

    # test
    for _ in range(1000):
        run(a_detach, b)
    after = time.perf_counter()
    t_req_grad_false = after - before

    print_table_row(B, M, float(t_req_grad_true), float(t_req_grad_false))
    # print("results allclose: ", torch.allclose(run(a, b), run(a_detach, b)))
    # print("max rel diff: ", ((run(a, b) -  run(a_detach, b)) / (run(a, b))).abs().max())


print_table_header()
measure(2048, 4)
measure(1024, 8)
measure(512, 64)
measure(128, 128)
measure(64, 512)
measure(8, 1024)
measure(4, 2048)
```

</details>


This means that when doing nn.Linear and if your batch size is low, you'll get 2x-3x improvements on that matmul during eval (e.g. no need materialize the large tensor during bw) simply by setting requires_grad=False. (With the PR that caused this regression, it also made it so that running eval with no_grad gives similar benefits for these input sizes)

What is also interesting is that running with compile with max-autotune does NOT help things today! In PT2, we always decompose via the composite implicit kernels, so by the time inductor sees the graph it will either contain mm + copy or bmm (rather than matmul), and inductor today isn't smart enough to transform from one strategy to the other.

**Comment by xuzhao9** [comment]
@soulitzer I use the following command to reproduce:

```
python run.py hf_Longformer -d cuda -t eval
```

**Comment by soulitzer** [comment]
Hmm I still don't observe a significant difference, GPU Time per batch was 41.47ms and 41.87ms respectively.

**Comment by lezcano** [comment]
> inductor today isn't smart enough to transform from one strategy to the other.

That's my point above. We should add this transformation to the set of transformations inductor tries.

**Comment by xuzhao9** [comment]
@soulitzer Can you please try again by adding the `--bs 4` parameter? In our CI environment, we use batch size 4 to saturate the A100 GPU.

**Comment by soulitzer** [comment]
Ah thanks @xuzhao9 that was it

**Comment by soulitzer** [comment]
> That's my point above. We should add this transformation to the set of transformations inductor tries.

Ah I misunderstood your original point. 

Yeah, it would definitely be good if inductor is able to handle this in some way. And one way is if during aot autograd we can choose not to run certain composite implicit kernels in the first place, so that inductor can see the original matmul, vs a bunch of transpose + clone + mm. But not sure how feasible that is, but on the surface it would make inductor's job a little easier. 
cc @bdhirsh

**Comment by soulitzer** [comment]
Creating a new issue to track the PT2 discussion since this one is closed
